### ISSUE TITLE
equalTo does not handle recursions involving global state correctly.

### ISSUE THREAD

**Issue by strangedev** [thenativeweb/assertthat#346]
In a scenario similar to #256 we were testing equality on some `HTMLElement`s generated by jsdom.
We either exceed the call stack size or memory limits. I had a hunch that this occurs because of recursions that use the global scope in some capacity.

I did some debugging and found a minimal example that exhibits this behavior:
```ts
import { assert } from './lib';

declare const someGlobalObject: any;

const globalScope = global as any;

const someLocalObject = {
  scope: 'local'
} as any;

const someOtherLocalObject = {
  scope: 'local',
  other: someLocalObject
};

someLocalObject.other = someOtherLocalObject;
assert.that(someLocalObject).is.equalTo(someOtherLocalObject);

globalScope.someGlobalObject = {
  scope: 'global',
  other: someLocalObject
};
someGlobalObject.self = someGlobalObject;
assert.that(someGlobalObject).is.equalTo(someGlobalObject);

someLocalObject.other = someGlobalObject;
assert.that(someLocalObject).is.equalTo(someGlobalObject);

```

What's interesting about this is, that only the last call fails. This leads me to believe that the recursion detection works in principle when using the global scope, but there's some edge case when it comes to mutual recursion.

**Comment by strangedev** [comment]
After some more digging I traced the error to https://github.com/thenativeweb/assertthat/blob/261b6d9ea153c1d09c917916e1ccb270c37f9b7b/lib/prettyPrint/forObjects/prettyPrintObject.ts#L15 It seems to me that the recursion is not picked up and the library tries to pretty print the objects, running into an infinite loop.

**Comment by strangedev** [comment]
I found the source of the error. The recursion is detected, but the `path` is `/` in both cases, tricking the comparison into thinking the `other` property is the same on both objects. This also happens with the non-global object. so these two are considered equal:

```ts
const someLocalObject = {
  scope: 'local'
} as any;

const someOtherLocalObject = {
  scope: 'local',
  other: someLocalObject
};

someLocalObject.other = someOtherLocalObject;
assert.that(someLocalObject).is.equalTo(someOtherLocalObject);
 ```

When the objects are considered to be equal, no diff is printed, hence no error.
If we change the `scope` property to something else, the library will try to pretty print the objects and run into an infinite loop:

```ts
const someLocalObject = {
  scope: 'local'
} as any;

const someOtherLocalObject = {
  scope: 'local as well',
  other: someLocalObject
};

someLocalObject.other = someOtherLocalObject;
assert.that(someLocalObject).is.equalTo(someOtherLocalObject);
```
I remember that we discussed this behavior in the past and settled on considering objects equal if they've got recursions in the same places that point to the same place structurally. So I think it is correct to set the recursion path to `/`.

We should, however, not pretty print the original objects without dispelling them first.
